### ISSUE TITLE
URL fetching with GitHub Releases

### ISSUE THREAD

**Issue by ax3l** [spack/spack#5356]
Somehow the url fetching of GitHub releases seems to mismatch the correct download if additional packages are uploaded.

e.g. the `protobuf` package should always fetch the `archive/vX.Y.Z.tar.gz` *github default packages* but a `spack checksum protobuf` also seems to randomly [catch other, additional uploads of a version](https://github.com/google/protobuf/releases):
```
==> Fetching https://github.com/google/protobuf/archive/v3.4.0rc3.tar.gz
######################################################################## 100,0%
==> Fetching https://github.com/google/protobuf/archive/v3.4.0rc2.tar.gz
######################################################################## 100,0%
==> Fetching https://github.com/google/protobuf/archive/v3.4.0rc1.tar.gz
######################################################################## 100,0%
==> Fetching https://github.com/google/protobuf/releases/download/v3.4.0/protobuf-cpp-3.4.0.tar.gz
######################################################################## 100,0%
==> Fetching https://github.com/google/protobuf/archive/v3.3.2.tar.gz
######################################################################## 100,0%
==> Fetching https://github.com/google/protobuf/archive/v3.3.1.tar.gz
######################################################################## 100,0%
==> Fetching https://github.com/google/protobuf/archive/v3.3.0rc1.tar.gz
######################################################################## 100,0%
==> Fetching https://github.com/google/protobuf/releases/download/v3.3.0/protobuf-ruby-3.3.0.tar.gz
######################################################################## 100,0%
==> Fetching https://github.com/google/protobuf/archive/v3.2.1.tar.gz
######################################################################## 100,0%
==> Fetching https://github.com/google/protobuf/releases/download/v3.2.0/protobuf-python-3.2.0.tar.gz
######################################################################## 100,0%
==> Checksummed 10 versions of protobuf

    version('3.4.0rc3', '51d32d9dffef91fa32ff5d2873e4bb96')
    version('3.4.0rc2', 'bb8d253a72f510603e050bdd87be1981')
    version('3.4.0rc1', '9ad0f36137cdb7cda605ca04a825310a')
    version('3.4.0',    '6d59dad503bea5ad420fd09ddad84481')
    version('3.3.2',    'ef2a6a6bb3b92d8fa8d71e3cef741f2f')
    version('3.3.1',    'a73549d2ac5f16a4027eecb17c28277f')
    version('3.3.0rc1', '2323e732953835a524680b6ede733a69')
    version('3.3.0',    '8b64807bc1fb52059873a31133ed71b2')
    version('3.2.1',    '23f80424b39dd1fff52ab84074380f87')
    version('3.2.0',    'ddc8f79ab4a1a6ade2c337fa55310ff4')
```

See for example the `3.2.0`, `3.3.0` and the `3.4.0` version: they should just use the simple
```
url      = "https://github.com/google/protobuf/archive/v3.2.0.tar.gz"
```
given in the `package.py`. Instead, it randomly takes one of the additional release artifacts.

Am I doing something wrong here or became fetching an url really that non-deterministic?

**Comment by ax3l** [comment]
@adamjstewart do you have an idea what could be wrong here?

**Comment by adamjstewart** [comment]
This is a fun one. In general, URL fetching on GitHub doesn't work the way most people think it does.

> they should just use the simple
> ```python
> url      = "https://github.com/google/protobuf/archive/v3.2.0.tar.gz"
> ```
> given in the `package.py`.

The way Spack's URL detection works is by going to the basename of the URL and looking for tarballs that are similarly named. What happens when you go to:
https://github.com/google/protobuf/archive ?
You get a 404 error. Spack realizes that this will always be the case, and GitHub is a common repository, so it has special logic to account for this that doesn't require people to manually add a `list_url` for every package. Check out the [`find_list_url` method in `url.py`](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/url.py#L65). You'll notice that for GitHub URLs, we automatically add a `list_url` of https://github.com/google/protobuf/releases. This page contains all of the links you see above.

Basically, this is the intended behavior.

**Comment by ax3l** [comment]
@adamjstewart all right, I understand that and even implemented the gitlab links in the past :)

but my question now is: how do I fix protobuf? I really want to always have the simple tagged tarballs and not a random one of the additional uploads o.0 :)

**Comment by adamjstewart** [comment]
There is currently no way to prevent this. If any webpage has multiple tarballs for the same version of the software, Spack cannot differentiate between the two.

**Comment by ax3l** [comment]
Ok, I will then overwrite protobuf's `url_for_version` for a fix

**Comment by ax3l** [comment]
Huh, my overwriting gets ignored and it still takes the other urls o.0

```diff
diff --git a/var/spack/repos/builtin/packages/protobuf/package.py b/var/spack/repos/builtin/packages/protobuf/package.py
index 0628974fd..0d2a25be5 100644
--- a/var/spack/repos/builtin/packages/protobuf/package.py
+++ b/var/spack/repos/builtin/packages/protobuf/package.py
@@ -46,6 +46,11 @@ class Protobuf(CMakePackage):
     # first fixed in 3.4.0: https://github.com/google/protobuf/pull/3406
     patch('pkgconfig.patch', when='@:3.3.2')
 
+    def url_for_version(self, version):
+        """fix for https://github.com/LLNL/spack/issues/5356"""
+        return ("https://github.com/google/protobuf/archive/"
+                "v{}.tar.gz".format(version.dotted))
+
     def cmake_args(self):
         args = [
             '-Dprotobuf_BUILD_TESTS:BOOL=OFF',
```

**Comment by adamjstewart** [comment]
`spack checksum` doesn't use `url_for_version`. It is used for fetching, not for finding new versions.

**Comment by scheibelp** [comment]
It looks like if you specify `spack checksum --versions=... protobuf` it will use `url_for_version`. Otherwise it will automatically add in the versions under `releases/` unless you were to override `Package.fetch_remote_versions()`. Is that reasonable? It seems like it isn't common for there to be unwanted archives under `releases/`.

**Comment by ax3l** [comment]
Isn't it confusing that the version fetcher and the checksum's version fetcher use different scraping mechanisms for finding a version?

> It seems like it isn't common for there to be unwanted archives under releases/.

yes, it's also the first package I came across that has that issue. We might need to keep in mind that especially github/gitlab *releases* can have arbitrary *additional artifacts* uploaded. most of the time people us those for binary artifacts, but in general it can be anything... as in the protobuf case additional `.tar.gz` files...

My proposal would still be to make version fetching deterministic first, so at least it's not randomly changing *which of the candidates is chosen* between subsequent calls :)

**Comment by adamjstewart** [comment]
> Isn't it confusing that the (install) version fetcher and the checksum's version fetcher use different scraping mechanisms/criteria for finding a version?

There is a reason for this. Let's take the simple case where all versions of a package live under the following structure:

https://www.example.com/releases/package-version.tar.gz

If https://www.example.com/releases/ is not listable, it is impossible for Spack to search for versions of the package. There may be a main website which has a link to the latest release, but Spack can't substitute the version with all possible version numbers and blindly search for a URL that works.

However, if the package has links to new versions on:

https://www.example.com/downloads/

`spack checksum` can search this page for links. The fetcher used by `spack fetch` and `spack install` is `url_for_version`, which requires knowledge of the URL structure AND the version being fetched. On the other hand, the fetcher used by `spack checksum` is `find_versions_of_archive`, which only requires knowledge of the webpage that lists available versions.

> My proposal would still be to make version fetching deterministic first

Yes, this would definitely be nice. I'm worried that I may have broken this in #2972, specifically my changes to `wildcard_version`. These changes allowed `spack checksum` to find significantly more versions for a lot of packages, but it also allowed Spack to match duplicate URLs for a single version. If you are going to play around with things, this would be a good place to start.

**Comment by adamjstewart** [comment]
Just checked and this is still broken on develop. If you remove the `fetch_remote_versions` hack you added to the package, the issue still exists, and likely affects far more packages than just `protobuf`.

**Comment by ax3l** [comment]
Hi @adamjstewart, I think this is solved by now with the updated URL schemes, right?
